Camera limiter class

hanbollar opened the pull request:
Fixes #4802

- [x]  figure out why camera flips between neg and pos values with _limitPosition and _limitOrientation
- [x]  _limitOrientation flips camera sideways when moving sometimes
- [x] good way to limit by "height" using sphereBBox - what if sphere not centered at origin - convert height to distance from center
- [x]  remove comments after finished debugging
- [x]  make sandcastle example (include limit by hpr and each of bounding objects)
- [x]  complete limiter spec
- [x]  add in camera tests with limiter included
- [x]  make sure Changes.md is updated
- [x] update for 2d
- [x] separate out hpr into different branch to be updated later - this now just has bounding object limiting locations

cesium-concierge commented:
![Signed CLA is on file.](https://img.shields.io/badge/CLA-signed-brightgreen.svg)

@hanbollar, thanks for the pull request! Maintainers, we have a signed CLA from @hanbollar, so you can review this at any time.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

hanbollar commented on Source/Scene/Camera.js:
```diff
@@ -980,10 +989,13 @@ define([
         updateMembers(this);
     };
 
-    var scratchAdjustOrtghographicFrustumMousePosition = new Cartesian2();
```
changed spelling here bc was spelled incorrectly - also updated only other time it was used in line 1011/1023

hanbollar commented on Source/Scene/Camera.js:
```diff
@@ -1134,6 +1146,10 @@ define([
     var scratchToHPRRight = new Cartesian3();
 
     function directionUpToHeadingPitchRoll(camera, position, orientation, result) {
+        if (!defined(result)) {
```
this if defined should be removed - my mistake - bc this method does not return anything

hanbollar commented on Source/Scene/Camera.js:
```diff
@@ -1687,6 +1708,9 @@ define([
         Matrix3.multiplyByVector(rotation, direction, direction);
         Matrix3.multiplyByVector(rotation, up, up);
         Matrix3.multiplyByVector(rotation, right, right);
+
```
should move this._limitPosition from 1510 to here

hanbollar replied:
was looking at wrong line - resolved this.

hanbollar commented on Source/Scene/Camera.js:
```diff
@@ -2806,6 +2835,7 @@ define([
         var orientation = defaultValue(options.orientation, defaultValue.EMPTY_OBJECT);
         if (defined(orientation.direction)) {
             orientation = directionUpToHeadingPitchRoll(this, destination, orientation, scratchSetViewOptions.orientation);
+            orientation = (defined(this.cameraLimiter)) ? this.cameraLimiter.closestOrientationTo(orientation) : orientation;
```
change this line to using this._limitOrientation - cleaner call

hanbollar replied:
why is this a one-liner but the method this._limitOrientation more than that?

hanbollar replied:
method is one liner. resolved.

hanbollar commented on Source/Scene/Camera.js:
```diff
@@ -2827,6 +2857,7 @@ define([
         if (isRectangle) {
             destination = this.getRectangleCameraCoordinates(destination, scratchFlyToDestination);
         }
+        destination = (defined(this.cameraLimiter)) ? this.cameraLimiter.closestLocationTo(destination) : destination;
```
change this line to using this._limitLocation - cleaner call

hanbollar commented on Source/Scene/Camera.js:
```diff
@@ -3243,5 +3274,60 @@ define([
+     * @private
+     */
+    Camera.prototype._limitPosition = function() {
+        // console.log('prev position:');
```
dont forget to remove these comments and rest of comments before done

hanbollar commented on Specs/Core/CameraLimiterSpec.js:
```diff
@@ -0,0 +1,71 @@
+//     //  // // WITHIN BOUNDING OBJECT -
+//     //
+//     //
+//     //  // WITHIN HEADING PITCH ROLL LIMITS -
```
add specs

hanbollar commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,301 @@
+        if (this.boundingObject.halfAxes.equals(Matrix3.ZERO)) {
+            return (position.equals(Cartesian3.ZERO));
+        }
+
```
@ggetz - could you review my math for here(_closestLocationInOrientedBoundingBox)

for closestLocationInOrientedBoundingBox what i'm currently doing is, since the halfAxes is scale of the item divided by two - to get the actual transformation, i'm scaling the overall by 2, using the inverse of this to transform back to object space (ie just a unit cube centered at the origin), calculating the appropriate min/maxes based on cube space, then converting back to worldspace for returned position.

for closestLocationInBoundingSphere -[my comments for me]- I'm "zeroing" the sphere by subtracting the center from the position so it's as if the position is relative to a sphere centered at zero - it seems i've currently forgotten to include a var storing this original centered location so that needs to be used instead of center for distToLocOnSphere, LocOnSphereInDir, and distToPosition.

hanbollar commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,301 @@
+        // var cartographicRep = Cartographic.fromCartesian(position);
+        // if (cartographicRep.height > )
+
+        var direction = Cartesian3.subtract(center, position, new Cartesian3());
```
this should be position - center not center - position

hanbollar commented on Source/Scene/Camera.js:
```diff
@@ -3243,5 +3274,60 @@ define([
+                direction : this.direction,
+                up : this.up,
+                right : this.right
+            }
```
needs semicolon

hanbollar commented on Source/Scene/Camera.js:
```diff
@@ -3243,5 +3274,60 @@ define([
+            // console.log('orientation changed');
+            // console.log(this.orientation);
+        }
+    }
```
needs semicolon

hanbollar commented on Source/Scene/Camera.js:
```diff
@@ -3243,5 +3274,60 @@ define([
+        //     console.log('position changed');
+        //     console.log(this.position);
+        // }
+    }
```
add semicolon

ggetz commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,301 @@
+            //>>includeStart('debug', pragmas.debug);
+            throw new DeveloperError('bounding object is not of an allowed type');
+            //>>includeEnd('debug');
+        }
```
I think these all have prototype clone methods. This can be shortened to

```
if (defined(limiter.boundingObject)) {
    limiter.boundingObject.clone(result);
}
```

and there is no need to throw the developer error.

ggetz commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,301 @@
+    CameraLimiter.clone = function(limiter, result) {
+        if (!defined(result)) {
+            var options = {};
+            result = new CameraLimiter(options);
```
No need for the options parameter, you already set the options to a default value in the constructor.

ggetz commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,301 @@
+            //>>includeEnd('debug');
+        }
+
+        // if min is defined, max must also be. if min is not defined, max must also not be.
```
Are we sure this is the case? Either way, we should not be throwing the developer error here.

ggetz commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,301 @@
+        var distanceToLocOnSphere = Cartesian3.distanceSquared(locationOnSphereInDirection, center);
+        var distanceToPosition = Cartesian3.distanceSquared(position, center);
+
+        return (distanceToPosition < distanceToLocOnSphere || distanceToPosition) ? position : locationOnSphereInDirection;
```
Can you find out if it's inside the sphere earlier and return the original position earlier in the function? That would avoid unnecessary operations and make this more readable as well.

ggetz commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,301 @@
+            cartesianPosition = this._closestLocationInOrientedBoundingBox(cartesianPosition);
+        } else {
+            //>>includeStart('debug', pragmas.debug);
+            throw new DeveloperError('Bounding Object not of allowed type. '
```
Maybe consider moving the `getClosestBoundedPosition` to each of these classes, and just calling `this.boundingObject.getClosestBoundedPosition() instead of all these cases.

pjcozzi commented:
Just a question on naming: are we use "Limiter" is the common term used here?  Is there a term from game or physics engines that we should consider?

ggetz commented:
> Just a question on naming: are we use "Limiter" is the common term used here? Is there a term from game or physics engines that we should consider?

The issue initially said "camera boundary", but as far as I know there is no common physics/game engine term. I'd probably lean towards "boundary".

mramato commented:
I think the term limit/limitter came from from the NASA WorldWind API (but I could be wrong; it's been forever).  Unless there is a standard term for this, I like "limit" over "boundary", because boundary connotes a geospatial area, but limits can also be applies to orientation and other properties.

pjcozzi commented:
Just please consider a 5-minute review of a representative game and psychics engine to confirm or see that it isn't common and stick with "limiter."

hpinkos commented on Source/Core/Cartographic.js:
```diff
@@ -117,6 +117,10 @@ define([
      * @returns {Cartographic} The modified result parameter, new Cartographic instance if none was provided, or undefined if the cartesian is at the center of the ellipsoid.
      */
     Cartographic.fromCartesian = function(cartesian, ellipsoid, result) {
+        if (!defined(result)) {
```
This isn't needed, it's handled below on L144

hpinkos commented on Source/Core/AxisAlignedBoundingBox.js:
```diff
@@ -191,6 +193,18 @@ define([
+     */
+    AxisAlignedBoundingBox.closestLocationIn = function(position, boundingObject) {
+        if (defined(boundingObject)) {
+            position.x = CesiumMath.clamp(position.x, boundingObject.minimum.x, boundingObject.maximum.x);
```
I don't think this is right.  `position` is in world space, and `boundingObject.minimum`/`boundingObject.maximum` are relative to `boundingObject.center`

hpinkos commented on Source/Core/BoundingRectangle.js:
```diff
@@ -318,6 +320,23 @@ define([
+    /**
+     * @private
+     */
+    BoundingRectangle.closestLocationIn = function(position, boundingObject) {
```
`BoundingRectangle` looks like it's in screen space, so I'm not sure this makes sense for a camera limiter

hpinkos commented on Source/Core/BoundingSphere.js:
```diff
@@ -1198,6 +1198,33 @@ define([
+            Cartesian3.normalize(direction, direction);
+            Cartesian3.multiplyByScalar(direction, radius, direction);
+
+            position = Cartesian3.add(direction, center, new Cartesian3());
```
Have the functions take a `result` parameter so we're not creating a `new Cartesian3()`

hpinkos commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,148 @@
+         * @type {HeadingPitchRoll}
+         * @default undefined
+         */
+        this.minHeadingPitchRoll = HeadingPitchRoll.clone(options.minimumHeadingPitchRoll);
```
What if I wanted to limit the `heading`, but not the `pitch` and `roll`?  Is there a way to do that?

hanbollar replied:
@hpinkos To do this a user would just set whichever minheading/pitch/roll maxheading/pitch/roll quantity they wished to be restricted and allow the rest to be Infinity/negInfinity

hpinkos commented on Source/Core/AxisAlignedBoundingBox.js:
```diff
@@ -191,6 +193,18 @@ define([
+    /**
+     * @private
+     */
+    AxisAlignedBoundingBox.closestLocationIn = function(position, boundingObject) {
```
I would probably rename these functions as `projectPoint` or something along those lines.  We are essentially projecting the position onto the bounding volume, right?

hpinkos commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -419,6 +419,34 @@ define([
+            var minimum = new Cartesian3(-1.0, -1.0, -1.0);
+            var maximum = new Cartesian3(1.0, 1.0, 1.0);
+
+            positionInObjectSpace.x = CesiumMath.clamp(positionInObjectSpace.x, minimum.x, maximum.x);
```
Same comment as for `AxisAlignedBoundingBox`.  `boundingObject.center` should be used somewhere in this function.

hpinkos commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -419,6 +419,34 @@ define([
+            }
+
+            // convert world space positionToCheck to orientedBoundingBox's object space.
+            var inverseTransformationMatrix = Matrix3.inverse(boundingObject.halfAxes, new Matrix3.fromScale(1.0));
```
`new Matrix3.fromScale(1.0)` -> `Matrix3.fromScale`

Also, functions like this are creating a `new Matrix3` if you don't define the result parameter.  For cases like this, we usually define a scratch variable outside the function scope.  Here is an example: 

https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/PolygonGeometry.js#L60
https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/PolygonGeometry.js#L64

Doing things like this prevent Cesium from slowing down from having to garbage collect a bunch of objects.

hpinkos commented on Source/Scene/Camera.js:
```diff
@@ -3243,5 +3259,27 @@ define([
      * @callback Camera~FlightCancelledCallback
      */
 
+    function _limitOrientation(angle, camera) {
```
You don't need to use an underscore for the name here.  That would only be if the function was on the object, ie `Camera.prototype._privateFunction` or `Camera._privateFunction`

hpinkos commented on Source/Scene/Camera.js:
```diff
@@ -1687,6 +1698,9 @@ define([
         Matrix3.multiplyByVector(rotation, up, up);
         Matrix3.multiplyByVector(rotation, right, right);
+
+        _limitOrientation(angle, this);
```
I'm not sure this is doing the right thing.  `angle` is the angle to move the camera by, and is not directly related to the heading/pitch/roll the camera is limited to

hpinkos commented on Source/Scene/Camera.js:
```diff
@@ -1743,6 +1757,8 @@ define([
         Cartesian3.cross(this.direction, this.up, this.right);
         Cartesian3.cross(this.right, this.direction, this.up);
 
+        _limitOrientation(angle, this);
```
same comment

hpinkos commented on Source/Scene/Camera.js:
```diff
@@ -3243,5 +3259,27 @@ define([
 
+    function _limitOrientation(angle, camera) {
+        var minCheck = (angle < 0);
+        var maxCheck = (angle > 0);
```
What if `angle === 0`?  We want to do something for that case, right?

hpinkos commented:
@hanbollar Super excited for this!  Our users have been asking for a feature like this for a while, so they will be thrilled that you're adding it.

ggetz commented:
@hanbollar I think you will be able to just call you limiter function at one place:  https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/Camera.js#L955 in `Camera.update`, as it's called before any frame is rendered.

hanbollar commented:
@ggetz Tutorial is up for review - lmk if you think this is appropriate - also im not too sure of great limiter values for the hpr so if you have recs for that that would be great! also lmk if outline visual of bounds is good idea or not.

ggetz commented:
@hanbollar Good start on the sandcastle example, here's my feedback:
- The visual of the bounds is good! Give the user a button to toggle being constrained or not rather than the dropdown. I'd like to be able to back the camera up and see the geometry I'm constrained by. 
- In your limit position function, make sure you are keeping the orientation the same. Checkout [`_adjustHeightForterrain`](https://github.com/hanbollar/cesium/blob/341f7e5a40bc87665762caec554d6ce6e3c3833f/Source/Scene/Camera.js#L370), it's doing something similar.
- Make the bounding sphere bigger, I often get stuck under terrain as it's loading in
- No abbreviated variable names
- Provide example of how most people will use this, setting the options when constructing:
```
viewer.camera.cameraLimiter = new Cesium.CameraLimiter({
    boundingObject: boundingSphere
});
```
- HPR is in radians, not degrees. I'd probably set them to things like `-Math.PI` to `Math.PI`
- Show example of using `Infinity` or `-Infinity` for options
- When HPR contrainsts are on,  sometimes I get the following error:
```
DeveloperError: normalized result is not a number
Error
    at new DeveloperError (http://localhost:8080/Source/Core/DeveloperError.js:43:19)
    at Function.Cartesian3.normalize (http://localhost:8080/Source/Core/Cartesian3.js:421:19)
    at rotateVertical (http://localhost:8080/Source/Scene/Camera.js:1794:28)
    at Camera.rotateUp (http://localhost:8080/Source/Scene/Camera.js:1785:9)
    at rotate3D (http://localhost:8080/Source/Scene/ScreenSpaceCameraController.js:1426:20)
    at tilt3DOnEllipsoid (http://localhost:8080/Source/Scene/ScreenSpaceCameraController.js:1665:9)
    at tilt3D (http://localhost:8080/Source/Scene/ScreenSpaceCameraController.js:1607:13)
    at reactToInput (http://localhost:8080/Source/Scene/ScreenSpaceCameraController.js:418:21)
    at update3D (http://localhost:8080/Source/Scene/ScreenSpaceCameraController.js:1938:9)
    at ScreenSpaceCameraController.update (http://localhost:8080/Source/Scene/ScreenSpaceCameraController.js:1971:13)
```

hanbollar commented:
@ggetz - made changes following your notes - currently the implementation does not allow for orientation/position changes in 2d mode and only allows position changes in Columbus view if the boundingObject is a BoundingSphere. limiter works for orientation and position in 3D.

still to do - specs for camera limiter and specs for camera with cameraLimiter in it.

lmk if the implementation so far works for what we discussed. Thanks!

ggetz commented on Apps/Sandcastle/gallery/CameraLimiter Tutorial.html:
```diff
@@ -0,0 +1,211 @@
+viewer.terrainProvider = new Cesium.CesiumTerrainProvider({
+    url : 'https://assets.agi.com/stk-terrain/v1/tilesets/world/tiles',
+    requestVertexNormals : true
+}); //-- causes black in 2d view if loaded and do zoom manipulations
```
Get rid of the comments here. Did you open an issue for the black when in 2d?

hanbollar replied:
it seemed to be caused by how my limiter was working. it shouldnt cause an issue anymore.

ggetz commented on Apps/Sandcastle/gallery/CameraLimiter Tutorial.html:
```diff
@@ -0,0 +1,211 @@
+// The Oriented Box
+var oriented = new Cesium.AxisAlignedBoundingBox(minForOriented, maxForOriented, centerForOriented);
+
+// View the Oriented Bounding Box - NEEED A BETTERRR WWAAAYYY TO DRAWWW (bbox is correct - visual isnt oriented the way it is.)
```
What the issue here?

hanbollar replied:
I dont think the oriented bounding box is actually rotated per the orientation of the box - i think the visual box that i have is actually just axis aligned (so it's not necessarily a good representation for this case but works well for the axis aligned version)

hanbollar commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,117 @@
+        if (defined(limiter.boundingObject)) {
+            result.boundingObject = limiter.boundingObject.clone(result.boundingObject);
+        }
+        result.minHeadingPitchRoll = HeadingPitchRoll.clone(limiter.minHeadingPitchRoll, result.minHeadingPitchRoll);
```
then HeadingPitchRoll will clone an undefined value into an undefined value per HeadingPitchRoll's own clone method. I'm assuming this is the appropriate clone output we wanted for that; however, let me know if you think I should change it to do different behavior.

hanbollar commented:
@ggetz thanks for the follow up! Just a few more questions regarding your previous comments - 

- In regards to the default hpr - the reason the default hpr limits flipped the globe upsidedown is that it was actually limiting the heading to match - the example should now include a better hpr limit that allows the initial values of the camera to exist - Ive also checked with different initial starting values of the hpr and they all do limit the hpr when set s.t. the camera's orientation is outside the initial limits. Could you show me an example where you're having the issue so I can better understand how it's not working for you?

- In regards to setting some of the HPR components to 0 resulting in the 'not normalized error' when moving the camera around. - i wasn't able to replicate this error - could you show me an example where it occurred for you?

as always - thanks for the help!

hanbollar commented:
@ggetz - resolved the normalized result error [if that was not the normalized result error you were talking about then ~resolved a new normalized result error~] 

I'm still not entirely sure what you mean when you're talking about the default hpr - could you explain this for me / demo it for me?

hanbollar commented:
@ggetz fixed hpr to be checked in local space instead of global space - i was able to replicate a limit on the camera's orientation - try it in sandcastle and see if it works for how you wanted

ggetz commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,132 @@
+        //>>includeStart('debug', pragmas.debug);
+        Check.typeOf.object('orientation', orientation);
+        Check.typeOf.object('result', result);
+        //>>includeEnd('debug');
```
You don't need to add these checks for private functions, same as above with `limitPosition`

ggetz commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,132 @@
+            result.pitch = Math.min(maxHPR.pitch, orientation.pitch);
+            result.roll = Math.min(maxHPR.roll, orientation.roll);
+        }
+        return result;
```
I'm still seeing some strange behavior for the Heading/Pitch/Roll limiting. Pitch is never limited, and heading only works when both min and max are set to `0.0` or `-/+infinity`.

I think the way this method should work would be to have the camera limiter take in the current orientation of the camera, and return the limited orientation. You could then take that value, and perform the matrix operations in the camera class itself in the `limitOrientation` function. Do you agree? We can then do all of the conversions from local to fixed frame there.

hanbollar replied:
just realized didnt convert back to global space for hpr after limiting the local version - also updated by putting the conversion in the camera class instead of in limithpr in the camera limiter class - up in next commit

hanbollar commented:
@ggetz lmk if you'd prefer different original hpr limits

hanbollar commented:
just as a note - switched from setView3D to setView bc setView3D uses Camera's original transformation which we did not update - still to resolve --> jitter issue?

ggetz commented:
@hanbollar In the interest of getting this merged, can we split the HPR limiting out into another branch until we figure out what's going on with properly updating the view? You should go ahead and finish up the tests with the bounding volumes too.

hanbollar commented:
@ggetz modified so just bounding object being used in the limiter - updated camera specs and created camera limiter specs to match - updated documentation in camera limiter as well as in camera to reflect this change.

lmk what you think and if this version without hpr is ready to be merged

ggetz commented on Apps/Sandcastle/gallery/Camera Limiter.html:
```diff
@@ -0,0 +1,195 @@
+<!DOCTYPE html>
```
The thumbnail just looks like the hello world example when in Sandcastle- Can we make it look different? Maybe show the boxes more closely. Also make sure the thumbnail is the same dimensions as the rest.

hanbollar replied:
checked dimensions and updated the new photo

ggetz commented on Apps/Sandcastle/gallery/Camera Limiter.html:
```diff
@@ -0,0 +1,195 @@
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, minimum-scale=1, user-scalable=no">
+    <meta name="description" content="Use the CameraLimiter to restrict certain aspects of the camera's movement.">
+    <meta name="cesium-sandcastle-labels" content="Tutorials, Showcases">
```
Not a `tutorial`.

ggetz commented on Apps/Sandcastle/gallery/Camera Limiter.html:
```diff
@@ -0,0 +1,195 @@
+    <meta charset="utf-8">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, minimum-scale=1, user-scalable=no">
+    <meta name="description" content="Use the CameraLimiter to restrict certain aspects of the camera's movement.">
```
Use the CameraLimiter to restrict ~certain aspects of~ the camera's movement

ggetz commented on Apps/Sandcastle/gallery/Camera Limiter.html:
```diff
@@ -0,0 +1,195 @@
+var viewer = new Cesium.Viewer('cesiumContainer');
+// store variables
+var scene = viewer.scene;
+// set up terrain
```
Here and throughout the example, no need for comments _this_ descriptive. There are other sandcastle examples that explain these. Does terrain need to be turned on though?

ggetz commented on Apps/Sandcastle/gallery/Camera Limiter.html:
```diff
@@ -0,0 +1,195 @@
+            viewer.camera.cameraLimiter.boundingObject = boundingSphere;
+        } else {
+            saveCameraLimiter.boundingObject = boundingSphere;
+        }
```
Why not just create one instance of a `cameraLimiter`, and set `viewer.camera.cameraLimiter` to that or undefined when enabling/disabling?

hanbollar replied:
this is so that when the user clicks on the drop down while the limiter is not on, when the limiter box is rechecked again - the bounding object the limiter is currently using actually reflects what is currently selected in the drop down

ggetz replied:
Doesn't this work?
```
var limtiter = new CameraLimiter(...);
```
```
onselect : function() {
    limiter.boundingObject = boundingSphere;
}
```
```
Sandcastle.addToggleButton('Limiter On', viewer.camera.cameraLimiter, function(checked) {
    if (checked) {
        viewer.camera.cameraLimiter = limiter;
    } else {
        viewer.camera.cameraLimiter = undefined;
    }
});
````

hanbollar replied:
that's partially the same idea of how i was implementing it - i can switch to that way for readability

ggetz commented on Apps/Sandcastle/gallery/Camera Limiter.html:
```diff
@@ -0,0 +1,195 @@
+});
+scene.globe.depthTestAgainstTerrain = true;
+
+
```
Here and throughout, get rid of the excess white space.

ggetz commented on CHANGES.md:
```diff
@@ -42,6 +42,8 @@ Change Log
 * Fixed bug where `AxisAlignedBoundingBox` did not copy over center value when cloning an undefined result. [#6183](https://github.com/AnalyticalGraphicsInc/cesium/pull/6183)
 * Fixed a bug where imagery stops loading when changing terrain in request render mode. [#6193](https://github.com/AnalyticalGraphicsInc/cesium/issues/6193)
 * Fixed `Resource.fetch` when called with no arguments [#6206](https://github.com/AnalyticalGraphicsInc/cesium/issues/6206)
+* Fixed discrepancy between default value used and commented value for default value for halfAxes of OrientedBoundingBox. [#6147](https://github.com/AnalyticalGraphicsInc/cesium/pull/6147)
```
Changes will now be in the April 2nd release. Make a new section if needed.

ggetz commented on Source/Core/AxisAlignedBoundingBox.js:
```diff
@@ -191,6 +193,27 @@ define([
+     * @param {Cartesian3} position The position being projected onto this AxisAlignedBoundingBox.
+     * @returns {Cartesian3} A projected version of the inputted position if it was not originally within the AxisAlignedBoundingBox.
+     */
+    AxisAlignedBoundingBox.prototype.projectedPoint = function(position, result) {
```
All of these new functions will need specs too.

ggetz replied:
Since the function will not modify the position if already inside the box, is `project` the best terminology? I think we may have touched on the naming before, so I'm sorry if we already came to a conclusion.

ggetz commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,70 @@
+    /**
+     * Duplicates this CameraLimiter instance.
+     *
+     * @param {CameraLimiter} [result] The object onto which to store the result.
```
missing the docs for the first param.

ggetz commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,70 @@
+     * The CameraLimiter is used to limit a {@link Camera}'s position (using a bounding object).
+     *
+     * If set, the camera of this limiter will be constrained by location.
+     * The camera ignores position limits in {@link SceneMode#SCENE2D} view and {@link SceneMode#COLUMBUS_VIEW}.
```
"Object used to define bounds limiting the camera's position when in 3D mode"?

Link to `Camera.cameraLimiter`. 

Remove:
> If set, the camera of this limiter will be constrained by location.
> The camera ignores position limits in {@link SceneMode#SCENE2D} view and {@link SceneMode#COLUMBUS_VIEW}.

That should be explained in `Camera.cameraLimiter`

ggetz commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,70 @@
+        options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+
+        /**
+         * The BoundingObject. Used to restrict camera positions of the {@link Camera} holding this CameraLimiter.
```
Don't just repeat the name of the variable, that's not very descriptive.

"The volume used to restrict a {@link Camera}'s position." ?

ggetz commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,70 @@
+        /**
+         * The BoundingObject. Used to restrict camera positions of the {@link Camera} holding this CameraLimiter.
+         * @type {AxisAlignedBoundingBox|BoundingSphere|OrientedBoundingBox}
+         * @default undefined
```
remove `@default undefined`, it's assumed unless otherwise specified.

ggetz commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,70 @@
+     * @alias CameraLimiter
+     * @constructor
+     *
+     * @param {AxisAlignedBoundingBox|BoundingSphere|OrientedBoundingBox} [options.boundingObject] Assigned to the boundingObject attribute which can be used to specify limits on locations.
```
Duplicate the description you use on `boundingObject` below.

hanbollar replied:
did this

ggetz replied:
This is still not the same as this line: https://github.com/AnalyticalGraphicsInc/cesium/pull/6172/files#diff-b1419f5d132d13190f35b5c255c34a0eR34

ggetz commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,70 @@
+        if (!defined(result)) {
+            result = new CameraLimiter();
+        }
+        if (!defined(limiter)) {
```
`limiter` should be a require parameter.

ggetz commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -419,6 +419,43 @@ define([
+     * @param {Cartesian3} position The position being projected onto this OrientedBoundingBox.
+     * @returns {Cartesian3} A projected version of the inputted position if it was not originally within the OrientedBoundingBox.
+     */
+    OrientedBoundingBox.prototype.projectedPoint = function(position, result) {
```
Don't forget the specs!

ggetz commented on Source/Scene/Camera.js:
```diff
@@ -1134,6 +1143,7 @@ define([
     var scratchToHPRRight = new Cartesian3();
 
     function directionUpToHeadingPitchRoll(camera, position, orientation, result) {
+
```
Remove extra space.

ggetz commented on Specs/Core/CameraLimiterSpec.js:
```diff
@@ -0,0 +1,142 @@
+
+     var scratchCartesian3 = new Cartesian3();
+     var secondScratchCartesian3 = new Cartesian3();
+     it('limited by AxisAligned bounding object with value outside of bounding object', function() {
```
Move the clamping logic to the tests for the `project` functions themselves. Here, just make sure we're calling the right function if the boundingObject is defined.

hanbollar commented:
@ggetz Thanks for all the feedback! lmk if there's anything else you think I should do

hanbollar commented:
@ggetz modified with your additional changes

ggetz commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,74 @@
+     *
+     * {@link @Camera#cameraLimiter}
+     *
+     * If set, the camera of this limiter will be constrained by location.
```
If what is set? Maybe `used by a {@link Camera} to limit the location of it's movement`?

ggetz commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,74 @@
+    /**
+     * Duplicates this CameraLimiter instance.
+     *
+     * @param {CameraLimiter} [limiter] The object being cloned.
```
`limiter` is required, remove the `[]`

ggetz commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,74 @@
+     */
+    CameraLimiter.clone = function(limiter, result) {
+        //>>includeStart('debug', pragmas.debug);
+        Check.typeOf.string(limiter, 'limiter');
```
I think this should be `Check.typeOf.object`, here and throughout.

ggetz commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,74 @@
+        options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+
+        /**
+         * The BoundingObject used to restrict camera positions of the {@link Camera} holding this CameraLimiter.
```
`the {@link Camera} holding this CameraLimiter.` -> `a {@link Camera}.`

ggetz commented on Source/Scene/Camera.js:
```diff
@@ -233,6 +235,13 @@ define([
         Cartesian3.multiplyByScalar(this.position, mag, this.position);
+
+        /**
+         * If set, the camera will have its movement constrained.
```
`the camera will have its movement constrained.` -> `the camera's position will be constrained.`

ggetz commented on Source/Scene/Camera.js:
```diff
@@ -3230,9 +3240,35 @@ define([
         result._transformChanged = true;
 
+        if (defined(camera.cameraLimiter)) {
+            CameraLimiter.clone(camera.cameraLimiter, result.cameraLimiter);
```
I don't think this is covered by the tests. You should be able to just add a line to the `clone` test.

hanbollar replied:
done.

ggetz commented on Specs/Core/BoundingSphereSpec.js:
```diff
@@ -913,5 +913,22 @@ defineSuite([
+        var sphere = new BoundingSphere(Cartesian3.ZERO, 3.0);
+        var testLocation = new Cartesian3(0, 0, 0);
+        var result = sphere.clampToBounds(testLocation, scratchCartesian3);
+
```
Add a few lines here for when the position != center for full code coverage.

ggetz commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -419,6 +419,42 @@ define([
+
+    /**
+     * If the given position is not already within the box, projects the given position onto the box.
+     * @param {Cartesian3} position The position being projected onto this OrientedBoundingBox.
```
Add the docs for the `result` parameter.

ggetz commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -419,6 +419,42 @@ define([
+        result = position.clone(result);
+        // to avoid inverse of a zero matrix (since that doesn't exist).
+        if (this.halfAxes.equals(Matrix3.ZERO)) {
+            result.x = 0.0;
```
This case is not covered by the specs.

hanbollar replied:
done.

ggetz commented on Source/Core/BoundingSphere.js:
```diff
@@ -1198,6 +1198,37 @@ define([
+    /**
+     * If the given position is not already within the sphere, projects the given position onto the sphere.
+     * @param {Cartesian3} position The position being projected onto this BoundingSphere.
+     * @returns {Cartesian3} A projected version of the inputted position if it was not originally within the BoundingSphere.
```
Add docs for `result` parameter

ggetz commented on Source/Core/AxisAlignedBoundingBox.js:
```diff
@@ -191,6 +193,27 @@ define([
+    /**
+     * If the given position is not already within the box, projects the given position onto the box
+     * @param {Cartesian3} position The position being projected onto this AxisAlignedBoundingBox.
+     * @returns {Cartesian3} A projected version of the inputted position if it was not originally within the AxisAlignedBoundingBox.
```
Add docs for `result` parameter

ggetz commented on Source/Core/BoundingSphere.js:
```diff
@@ -1198,6 +1198,37 @@ define([
+        Cartesian3.multiplyByScalar(result, radius, result);
+
+        Cartesian3.add(result, center, result);
+        return result;
```
Recently we updated the [coding guide on result parameters](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/CodingGuide#result-parameters-and-scratch-variables). (Last paragraph of that section specifically).

For clarity, we should have 

```
result = Cartesian3.subtract(result, center, result);
```

etc.

ggetz commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -419,6 +419,42 @@ define([
+     */
+    OrientedBoundingBox.prototype.clampToBounds = function(position, result) {
+        //>>includeStart('debug', pragmas.debug);
+        Check.defined('position', position);
```
Use `Check.typeof.Object`

hanbollar replied:
when i do that i claims the position is type of string thought it shows the [object object] statement for type of position - ideas?

ggetz replied:
Check the order of your parameters - I think you have them swapped.

ggetz commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -419,6 +419,42 @@ define([
         return result;
     };
 
+    var scratchMatrix3PP = new Matrix3();
```
Don't use abbreviations like `PP`. `scratchMatrix` is fine.

ggetz commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -419,6 +419,42 @@ define([
+        var inverseTransformationMatrix = Matrix3.inverse(this.halfAxes, scratchMatrix3PP);
+        var positionInObjectSpace = Matrix3.multiplyByVector(inverseTransformationMatrix, result, scratchCartesian3PP);
+
+        // once in object space just translate to center and check if converted location is within axis oriented 2x2x2 cube (bc half axes)
```
~(bc half axes)~

ggetz commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -419,6 +419,42 @@ define([
+        result.z = CesiumMath.clamp(result.z, -1.0, 1.0);
+        Cartesian3.add(result, this.center, result);
+
+        Matrix3.multiplyByVector(this.halfAxes, result, result);
```
I think the order of these last lines should be swapped.

Get the position in object space -> Multiply by halfAxes -> Add new relative vector to the center.

hanbollar commented:
@ggetz ready for review again

hpinkos commented on CHANGES.md:
```diff
@@ -61,7 +63,7 @@ Change Log
 * Fixed bug where `AxisAlignedBoundingBox` did not copy over center value when cloning an undefined result. [#6183](https://github.com/AnalyticalGraphicsInc/cesium/pull/6183)
 * Fixed a bug where imagery stops loading when changing terrain in request render mode. [#6193](https://github.com/AnalyticalGraphicsInc/cesium/issues/6193)
-* Fixed `Resource.fetch` when called with no arguments [#6206](https://github.com/AnalyticalGraphicsInc/cesium/issues/6206)
+* Fixed `Resource.fetch` when called with no arguments [#6206](https://github.com/AnalyticalGraphicsInc/cesium/issues/6206
```
You have some cleanup to do here.  None of these older entries should be changed

hpinkos commented on Source/Core/AxisAlignedBoundingBox.js:
```diff
@@ -191,6 +193,28 @@ define([
+        Check.typeOf.object('position', position);
+        //>>includeEnd('debug');
+
+        result = position.clone(result);
```
Use `result = Cartesian3.clone(position, result)` instead.  We always prefer to use the static function for math classes like `Cartesian3`.  See https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/CodingGuide#use-prototype-functions-for-fundamental-classes-sparingly

hpinkos replied:
Same throughout, you have this in a few different places

hpinkos commented on Source/Core/BoundingSphere.js:
```diff
@@ -1198,6 +1198,36 @@ define([
+     */
+    BoundingSphere.prototype.clampToBounds = function(position, result) {
+        //>>includeStart('debug', pragmas.debug);
+        console.log(position);
```
remove this

hpinkos commented on Source/Core/BoundingSphere.js:
```diff
@@ -1198,6 +1198,36 @@ define([
+        result = position.clone(result);
+
+        // to avoid dividing by zero check if same location as center
+        if (result.equals(this.center)) {
```
Same here. `Cartesian3.equals(result, this.center)`

hpinkos commented on Source/Core/BoundingSphere.js:
```diff
@@ -1198,6 +1198,36 @@ define([
+        //>>includeEnd('debug');
+        result = position.clone(result);
+
+        // to avoid dividing by zero check if same location as center
```
Don't use inline comments unless something is non-obvious

hpinkos replied:
Same throughout

hpinkos commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,73 @@
+     *
+     * @param {AxisAlignedBoundingBox|BoundingSphere|OrientedBoundingBox} [options.boundingObject] The BoundingObject used to restrict camera positions of the {@link Camera} holding this CameraLimiter.
+     */
+    function CameraLimiter(options) {
```
This sholdn't be an `options` parameter if there's only one.  We use an `options` object when a function takes a ton of different parameters to make it more concise.  See https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/CodingGuide#options-parameters

hpinkos commented on Source/Core/CameraLimiter.js:
```diff
@@ -0,0 +1,73 @@
+     * @alias CameraLimiter
+     * @constructor
+     *
+     * @param {AxisAlignedBoundingBox|BoundingSphere|OrientedBoundingBox} [options.boundingObject] The BoundingObject used to restrict camera positions of the {@link Camera} holding this CameraLimiter.
```
This sholdn't be an `options` parameter if there's only one.  We use an `options` object when a function takes a ton of different parameters to make it more concise.  See https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/CodingGuide#options-parameters

Also, is this parameter required?

hpinkos replied:
Do we need a separate class for this?  There's hardly any logic in this class, why not just put it all in `Camera`?

hanbollar replied:
originally this class contain hpr constraints and a bounding object - which is why the options parameter was needed - we're only splitting out into this branch [ie it just has bounding object instead of bobject and hpr] so we can try to merge with just it for now and remove the current jitter issue from the hpr constraint in another issue

hpinkos commented on Source/Scene/Camera.js:
```diff
@@ -3230,9 +3240,35 @@ define([
+    function limitPosition(camera) {
+        // following the convention from _adjustHeightForTerrain to maintain correct orientation
+        // before and after position is changed
+        if (defined(camera.cameraLimiter) && camera._mode === SceneMode.SCENE3D) {
```
Instead of putting everything in an `if` block, we prefer to return early.  ie
```js
if (!defined(camera.cameraLimiter) || camera._mode !== SceneMode.SCENE3D) {
    return;
}
// continue limiting logic here

hpinkos replied:
Why does this only work in `SCENE3D`?  I would expect this to work in all 3 scene modes

hanbollar replied:
the current implementation does not support other scene modes - this is an issue that still needs to be resolved

hpinkos replied:
Gotcha.  I think that needs to be resolved before this gets merged.  If you need help figuring it out, you can talk to me or @bagnell

hpinkos commented on Source/Scene/Camera.js:
```diff
@@ -3230,9 +3240,35 @@ define([
+        // before and after position is changed
+        if (defined(camera.cameraLimiter) && camera._mode === SceneMode.SCENE3D) {
+            // store camera's hpr values
+            saveOrientation.heading = camera.heading;
```
I'm not sure this is needed.  I think `setView` already uses the current camera orientation if it's not specified

hanbollar replied:
it uses the current camera direction - when i tested both versions in the camera limiter tutorial it seems that not saving the previous orientation does not allow for rotations that are not centered around the globe - should this be the case though? [also the wording here's kind of weird so i can demo it for u if needed]

hpinkos replied:
Okay, I took a look at the code and you're correct, `setView` does not maintain HPR.  I think it used to, but we changed it so the behavior was the same as `flyTo`

hpinkos commented on Source/Scene/Camera.js:
```diff
@@ -3230,9 +3240,35 @@ define([
+            saveOrientation.roll = camera.roll;
+
+            // update position
+            camera.position = camera.cameraLimiter.limitPosition(camera.position, camera.position);
```
Don't set the `camera.position` here.  Instead, just send the result to `setView`.  `setView` has logic that will make it work in all 3 scene modes

hpinkos commented on Specs/Core/AxisAlignedBoundingBoxSpec.js:
```diff
@@ -161,4 +161,30 @@ defineSuite([
+        var testLocation = new Cartesian3(3, 3, 3);
+        var result = box.clampToBounds(testLocation, scratchCartesian3);
+
+        expect(result).not.toEqual(testLocation);
```
A more precise test would be good here.  Checking that it doesn't equal the original value doesn't mean that it returned the correct value.

hanbollar replied:
done.

ggetz commented on Source/Scene/Camera.js:
```diff
@@ -3230,9 +3241,39 @@ define([
+
+    /**
+     * @private
+     */
```
This block comment is describing and should be above `Camera.prototype._limitPosition`, not `var scratchClampedValue`

ggetz commented on Source/Scene/Camera.js:
```diff
@@ -3230,9 +3241,39 @@ define([
         return result;
     };
 
+    var options = {
```
For clarity, let's call this 'scratchViewOptions`.

ggetz commented on Specs/Core/AxisAlignedBoundingBoxSpec.js:
```diff
@@ -161,4 +161,31 @@ defineSuite([
     });
+
+    var scratchCartesian3 = new Cartesian3();
+    var secondScratchCartesian3 = new Cartesian3();
```
In the future, don't worry about using scratch variables for tests, they're not needed.

ggetz commented on Specs/Core/BoundingSphereSpec.js:
```diff
@@ -913,5 +913,33 @@ defineSuite([
+
+        expect(result).not.toEqual(testLocation);
+
+        var result2 = sphere.clampToBounds(result, scratchCartesian3);
```
I don't think this is a very precise test. We should be able to figure out what the result is after clamping (`Cartesian3(3, 3,3)`, correct? So we should check that it is returning that value.

ggetz replied:
Same with the other bounding tests.

ggetz commented on Specs/Scene/CameraSpec.js:
```diff
@@ -3059,4 +3063,106 @@ defineSuite([
+
+        spyOn(camera.boundingObject, 'clampToBounds');
+        camera._limitPosition();
+        expect(camera.boundingObject.clampToBounds).toHaveBeenCalled();
```
Add a check for the new resulting `camera.position` as well please.

hpinkos commented:
@hanbollar I'll take one more look at this after you address @ggetz comments.

Before this is ready to merge, can you squash your commits?  It's not something we always do, but since you have so many commits in this PR I think it would be a good idea.

hanbollar commented:
@hpinkos and @ggetz finished recent edits and squashed as requested

ggetz commented:
Thanks @hanbollar, @hpinkos do you want to take the final look and merge when you're happy?

hpinkos commented on CHANGES.md:
```diff
@@ -31,6 +32,7 @@ Change Log
 * Fixed support of glTF-supplied tangent vectors. [#6302](https://github.com/AnalyticalGraphicsInc/cesium/pull/6302)
 * Fixed improper zoom during model load failure. [#6305](https://github.com/AnalyticalGraphicsInc/cesium/pull/6305)
 * Fixed model loading failure when containing unused materials. [6315](https://github.com/AnalyticalGraphicsInc/cesium/pull/6315)
+* Fixed discrepancy between default value used and commented value for default value for halfAxes of OrientedBoundingBox. [#6147](https://github.com/AnalyticalGraphicsInc/cesium/pull/6147)
```
Revert this and the below changes in `CHANGES.md`.  These look like they were the result of a bad merge somewhere along the way.

hanbollar replied:
confused about this - discuss with me when you get in

hpinkos commented on Source/Core/AxisAlignedBoundingBox.js:
```diff
@@ -191,6 +193,28 @@ define([
+     * @param {Cartesian3} [result] The object onto which to store the result.
+     * @returns {Cartesian3} A projected version of the inputted position if it was not originally within the AxisAlignedBoundingBox.
+     */
+    AxisAlignedBoundingBox.prototype.clampToBounds = function(position, result) {
```
I would require the `result` parameter.  We usually only make it optional for functions like `fromCartesian3`, `fromRectangle`, where you're creating a new instance

hpinkos replied:
Same for the other `clampToBounds` functions in the other bounding volume types

hpinkos commented on Source/Core/AxisAlignedBoundingBox.js:
```diff
@@ -191,6 +193,28 @@ define([
     };
 
+    /**
+     * If the given position is not already within the box, projects the given position onto the box
```
I don't think `projection` is the right term here.  This function does clamp the position within the bounds of the box.  `projection` makes me think that the result would be the intersection of the vector for the center of the box to the position with one of the box faces, and this doesn't do that.  I think we just want to update the doc here, this clamping behavior is what we want.

hpinkos commented on Source/Core/BoundingSphere.js:
```diff
@@ -1198,6 +1198,35 @@ define([
+        //>>includeEnd('debug');
+        result = Cartesian3.clone(position, result);
+
+        // to avoid dividing by zero
```
Throughout, remove inline comments like this.  We only use them if something is really non-obvious

hpinkos commented on Source/Core/OrientedBoundingBox.js:
```diff
@@ -419,6 +419,45 @@ define([
+        result = Cartesian3.clone(position, result);
+        // to avoid inverse of a zero matrix (since that doesn't exist).
+        if (this.halfAxes.equals(Matrix3.ZERO)) {
+            result.x = 0.0;
```
Wouldn't the result be the center of the box in this case?  This is a box with no dimension, but it still has a center position.

hpinkos commented on Source/Scene/Camera.js:
```diff
@@ -233,6 +237,12 @@ define([
         Cartesian3.multiplyByScalar(this.position, mag, this.position);
+
+        /**
+         * If set, the camera's position is constrained to be within this bounding object.
```
This doc isn't quite right.  It should be like `Gets and sets the bounding object used to constrain the camera's position`

hanbollar replied:
I'm not really sure I understand why this needs to be changed? it's not a getter or a setter itself?

hpinkos replied:
This is just the standard way we document properties like this. It lets the user know that they can read and write to this property.

hpinkos commented on Source/Scene/Camera.js:
```diff
@@ -233,6 +237,12 @@ define([
+
+        /**
+         * If set, the camera's position is constrained to be within this bounding object.
+         * @type {AxisAlignedBoundingBox|BoundingSphere|OrientedBoundingBox} [options.boundingObject] The BoundingObject used to restrict positions of the {@link Camera}.
```
Remove `[options.boundingObject]`

hpinkos replied:
`BoundingObject` -> `bounding object`

hpinkos commented on Source/Scene/Camera.js:
```diff
@@ -233,6 +237,12 @@ define([
+         * If set, the camera's position is constrained to be within this bounding object.
+         * @type {AxisAlignedBoundingBox|BoundingSphere|OrientedBoundingBox} [options.boundingObject] The BoundingObject used to restrict positions of the {@link Camera}.
+         */
+        this.boundingObject = undefined;
```
Move this so it's where the other public attributes are.  Around Line 215

hpinkos commented on Source/Scene/Camera.js:
```diff
@@ -3230,9 +3241,39 @@ define([
+        scratchViewOptions.orientation.roll = this.roll;
+
+        // solving for positionWC because there's a bug where WC is incorrect in 2D and CV mode
+        scratchViewOptions.destination = this.boundingObject.clampToBounds(Cartographic.toCartesian(this._positionCartographic), scratchClampedValue);
```
Using `Cartographic.toCartesian` make sit generate a `new Cartesian3()` every time this is called.
This is also not compatible with different ellipsoid types.  Use `this._scene.globe.ellipsoid.cartographicToCartesian` instead.

Also, at the top of the function you'll need to return if `scene.globe` is undefined

hpinkos commented:
This could really use a Sandcastle example.  You said in the first comment that you made one.  Did you forget to re-add it after you squashed your commits?

ggetz commented:
> This could really use a Sandcastle example. You said in the first comment that you made one. Did you forget to re-add it after you squashed your commits?

@hanbollar Definitely included an example initially, it should be a matter of putting it back.

hanbollar commented:
@hpinkos - my mistake - i added with . instead of * the first time so it missed all the new specs / tutorials etc. Updated per your recommended changes, but i did have some questions regarding your note for Changes.md and your first note for the Camera.js - lmk when you're free to discuss this

hanbollar commented:
@hpinkos @ggetz Thanks for all the feedback! lmk if there's anything else you think I should do

ggetz commented on Apps/Sandcastle/gallery/Camera Limiter.html:
```diff
@@ -0,0 +1,168 @@
+// View the Bounding Sphere
+var outlineSphere = viewer.entities.add({
+    name : 'Bounding Sphere Outline',
+    position: new Cesium.Cartesian3(0.01, 0.01, 0.01), // errors if 0,0,0
```
Can we set this somewhere else then? 

Also like @hpinkos said, in general just limit the comments to things that are non-obvious.

ggetz commented:
Closed in favor of https://github.com/AnalyticalGraphicsInc/cesium/pull/6381
